### ISSUE TITLE
Defining local scope variables in macros

### ISSUE THREAD

**Issue by okohub** [PebbleTemplates/pebble#371]
Hello,

When dealing with macros, sometimes we need to hold a value in variable to use it more than once.

For example;

```
{% macro toggleMacro(itemInfo) %}
{% set productId = itemInfo.item.productId %}
<input type="radio"
       class="select-radio hide"
       checked="true"
       value="{{ currentUnit }}"
       name="{{ 'unit_' + productId }}"/>
{% endmacro %}
```

but setting a new productId variable, will also override current productId variable if exists in context. I know it can be useful to reach all context, but I think we need to also support local variables.

May be {% set local foo = bar %} for example?

**Comment by ebussieres** [comment]
Macros have access to all variables within the template and are no longer restricted to a "local scope" in version 3.0.

Which version are you using ?

**Comment by okohub** [comment]
We are using version 3.0.0, but I want to make my context variables strict to current macro scope as I said before.

```
{% set productId = itemInfo.item.productId %}
```

this expression will override current variable if set outside macro, or parent template. But we don't want to do this.

Also, this isn't bit risky? If someone forgets a common variable name here like 'username' may be, finding out bug will be very hard to trace.

What do you think?

**Comment by cjbrooks12** [comment]
I'm not too familiar with the template scoping rules, but it looks like macros push a new scope into the ScopeChain, so variables set within the context of the macro should not affect the same-named variables outside the macro. Have you verified that setting the productId within the macro does persist outside of it?

Also, Twig has a [`with` tag](https://twig.symfony.com/doc/2.x/tags/with.html) which creates a new local scope on-demand, maybe this is something we could implement here @ebussieres? I'd be happy to make a PR for this

**Comment by okohub** [comment]
@cjbrooks12 I will provide an example project when I am available

**Comment by okohub** [comment]
@ebussieres @cjbrooks12 hello again,

https://github.com/okohub/pebbledemo

Here, this a sample project that demonstrates problem. A macro can change a variable that not defined in itself. This is risky.

**Comment by ebussieres** [comment]
I know why it's happening. Imagine the following case

```
{% set total = 0 %}
{% for item in items %}
  {% set total = total + item.balance %}
{% endfor %}
{{ total }}
```

If I made the requested change, this case won't work anymore. Set tag sets the value of a variable in the first scope in the chain that already contains the variable. Otherwise, it adds a variable to the current scope if an existing variable is not found.

And I think that setting a variable in your example is not necessary. You can change it to something like 

```
{% macro toggleMacro(itemInfo) %}
<input type="radio"
       class="select-radio hide"
       checked="true"
       value="{{ currentUnit }}"
       name="{{ 'unit_' + itemInfo.item.productId }}"/>
{% endmacro %}
```

**Comment by okohub** [comment]
My example is very basic as you see, there are more complex macros in our applications. Imagine when you are accessing 4th or 5th level field of objects multiple times, didn't you define a temp variable for this? 

I think you don't need to make a breaking change. A local scope for macros will make the trick. I like the idea what @cjbrooks12 suggests. An on demand local scope will be handy :)

**Comment by okohub** [comment]
I think the problem starts from that #231 th issue. 

As you can see there are another people who think this access is not necessary and risky. -> https://github.com/PebbleTemplates/pebble/issues/242

I understand, you should plan a new major change for this, but for now, creating dynamic local scope can handle this situation, we can not ignore this.

**Comment by ebussieres** [comment]
I also think that macro should only have a local scope. But we'll need to think about how to access global variable.

**Comment by ebussieres** [comment]
Resolved via #375

**Comment by SharkFourSix** [comment]
> Macros have access to all variables within the template and are no longer restricted to a "local scope" in version 3.0.
> 
> Which version are you using ?



> Macros have access to all variables within the template and are no longer restricted to a "local scope" in version 3.0.
> 
> Which version are you using ?

That's funny because I'm using 3.1.5 and I'm not getting the behavior that you are describing. Was something changed?
